Background image upload: Update `parent_id` of already uploaded images

selanthiraiyan opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #7021 

### Description
This PR is for updating the `parent_id`  of already uploaded images, using the remote product ID. (behind `backgroundProductImageUpload` feature flag)

- While creating a product ID we use a local product ID (which is zero) for uploading the product images. Once the product is created, the images that are already uploaded still have the `parent_id` set as 0. This PR updates the product ID of already uploaded images using [this API](https://developer.wordpress.com/docs/api/1.1/post/sites/%24site/media/%24media_ID/). 
- While creating a product ID we also update the product ID of the `ProductImageActionHandler` with the remote product ID to make sure that images that are uploaded after the product use the right product ID.

### Testing instructions

#### Feature flag on

- Launch the app
- Navigate to `Products` tab and tap on `+` button on top right.
- In product creation form enter a product title and add `image A`.
- Wait for `image A` to finish uploading.
- Tap "Publish" to save the product. 
- Stay on the product detail screen.
- Add another `image B` and wait for it to finish uploading and save the product again.
- From your computer, open your store's `Media Library` by navigating to https://yourstore.com/wp-admin/upload.php
- In the `Media Library` page,
  - Tap on `image A` and verify that `Uploaded to:` field is available. [Ref screenshot](https://user-images.githubusercontent.com/524475/175935720-22ba1b22-365b-49e5-95a1-59522dd9a58f.png) You can tap on the link to open the product that you created above.
  - Tap on `image B` and verify that `Uploaded to:` field is available for `image B` as well. 
  
#### Feature flag off

- Launch the app
- Navigate to `Products` tab and tap on `+` button on top right.
- In product creation form enter a product title and add `image A`.
- Wait for `image A` to finish uploading.
- Tap "Publish" to save the product. 
- Stay on the product detail screen.
- Add another `image B` and wait for it to finish uploading and save the product again.
- From your computer, open your store's `Media Library` by navigating to https://yourstore.com/wp-admin/upload.php
- In the `Media Library` page,
  - Tap on `image A` and verify that `Uploaded to:` field is not available.
  - Tap on `image B` and verify that `Uploaded to:` field is not available.
  

### Screenshots
**Add images following the steps in the video below.**
https://user-images.githubusercontent.com/524475/175932567-ec893530-9ec9-4a57-af98-9f747b9ff307.mov


**Check the images from`wp-admin`. Screenshots below**
![FlowerImage](https://user-images.githubusercontent.com/524475/175935720-22ba1b22-365b-49e5-95a1-59522dd9a58f.png)
<img width="1295" alt="Screenshot 2022-06-27 at 5 05 52 PM" src="https://user-images.githubusercontent.com/524475/175933232-1dbf4804-7517-4f3f-a38a-7c0caf6b0d08.png">



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Add a post method to update parent_id of media.

selanthiraiyan committed:
- Introduce `ProductImagesProductIDUpdater` for updating remote product ID to uploaded images.

selanthiraiyan committed:
Add a method to update product ID in `ProductImageActionHandler`.

selanthiraiyan committed:
Update `parent_id` of images uploaded with local productID.

selanthiraiyan committed:
Replace local product ID with remote product ID in `ProductImageActionHandler`.

selanthiraiyan committed:
Merge branch 'trunk' into feat/7021-update-parent_id-of-images

selanthiraiyan committed:
Add json for simulating update product id response.

selanthiraiyan committed:
Add `updateProductID` case to `MockMediaStoresManager`.

selanthiraiyan committed:
Network layer tests for Media `updateProductID` action.

selanthiraiyan committed:
Add `MediaStore` tests for `MediaAction.updateProductID`.

selanthiraiyan committed:
Refactor `ProductImagesProductIDUpdater` to use async await pattern.

selanthiraiyan committed:
Remove plural form from method name.

selanthiraiyan committed:
Rename method for clarity add documentation.

selanthiraiyan committed:
`ProductImagesProductIDUpdater`
- Add test to ensure `MediaAction.updateProductID` is dispatched to update `parent_id` of media.

selanthiraiyan committed:
Rename method for clarity and add comments.

selanthiraiyan committed:
Inject `ProductImagesProductIDUpdater` for testability.

selanthiraiyan committed:
Remove space.

selanthiraiyan committed:
Test `ProductImageUploader` fires `updateImageProductID` function in `ProductImagesProductIDUpdater`.

selanthiraiyan committed:
Fix whitespace violations.

selanthiraiyan committed:
Remove extra whitespace.

jaclync commented on Networking/Networking/Remote/MediaRemote.swift:
```diff
@@ -164,6 +168,31 @@ public class MediaRemote: Remote, MediaRemoteProtocol {
+            ParameterKey.fieldsWordPressSite: ParameterValue.wordPressMediaFields,
+        ]
+        let path = "sites/\(siteID)/media/\(mediaID)"
+        let request = DotcomRequest(wordpressApiVersion: .mark1_1, method: .post, path: path, parameters: formParameters)
```
Let's make sure this endpoint works for JCP sites (sites without Jetpack-the-plugin)

selanthiraiyan replied:
I created a store using jurassic ninja and installed WooCommerce Pay plugin to test this. 

The request fails with the following error,
```
{
  "error": "api_not_found",
  "message": "The Jetpack site is inaccessible or returned an error: server error. requested method jetpack.jsonAPI does not exist. [-32601]"
}
```
😢

selanthiraiyan replied:
I will take a look at using https://developer.wordpress.org/rest-api/reference/media/#update-a-media-item for JCP sites and update here.

jaclync replied:
I was just going to share the same after the meeting 😄 I can check it out today and share the result here!

jaclync replied:
Just tested the WP endpoint, and it worked! Here's the commit with the changes on the WP endpoint and in Yosemite: https://github.com/woocommerce/woocommerce-ios/commit/1f55b3a4ccc2c20a2ec70c8b5fed682c45d98b58

selanthiraiyan replied:
I have created a separate PR for covering the JCP sites. https://github.com/woocommerce/woocommerce-ios/pull/7167

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -193,4 +193,30 @@ final class ProductImageUploaderTests: XCTestCase {
+        /// `mockProductIDUpdater` should have received calls to update product ID of the already uploaded images.
+        waitUntil(condition: {
+            mockProductIDUpdater.numberOfTimesUpdateImageProductIDWasCalled == alreadyUploadedImages.count
+        })
```
nit: this might be simplified as:
```suggestion
        waitUntil {
            mockProductIDUpdater.numberOfTimesUpdateImageProductIDWasCalled == alreadyUploadedImages.count
        }
```

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/7152/commits/c2012a9ea169b34f0521c8e70f6fd5698baecdcd

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImagesProductIDUpdaterTests.swift:
```diff
@@ -0,0 +1,39 @@
+
+    override func setUp() {
+        super.setUp()
+        storesManager = MockStoresManager(sessionManager: SessionManager.testingInstance)
```
super nit: 
```suggestion
        storesManager = MockStoresManager(sessionManager:.testingInstance)
```

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/7152/commits/5592aa81ff42bac4a835032f5d68348df9c32a1b

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImagesProductIDUpdaterTests.swift:
```diff
@@ -0,0 +1,39 @@
+    override func setUp() {
+        super.setUp()
+        storesManager = MockStoresManager(sessionManager: SessionManager.testingInstance)
+        ServiceLocator.setStores(storesManager)
```
it looks like `ProductImagesProductIDUpdater` already takes in a `StoresManager`, do we still need to set it to the global singleton?

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/7152/commits/164e70ae132ee29767163feac11257fad1232fbc

jaclync commented on Yosemite/YosemiteTests/Stores/MediaStoreTests.swift:
```diff
@@ -416,6 +420,70 @@ final class MediaStoreTests: XCTestCase {
+    func test_updateProductID_returns_media() throws {
+        // Given
+        let path = "sites/\(sampleSiteID)/media/\(sampleMediaID)"
+        network.simulateResponse(requestUrlSuffix: path, filename: "media-update-product-id")
```
nit: maybe we can use `MockMediaRemote.whenUpdatingProductIDResultsBySiteIDToWordPressSite` in this and the next test case instead of the json response so that we can tweak its result in a more flexible way? Right now `whenUpdatingProductIDResultsBySiteIDToWordPressSite` doesn't seem used

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/7152/commits/e8fc7542465966df75a15771195cf54ffee4a1df

jaclync commented on WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift:
```diff
@@ -80,9 +83,17 @@ final class ProductImageUploader: ProductImageUploaderProtocol {
         let keyWithRemoteProductID = ProductKey(siteID: siteID, productID: remoteProductID, isLocalID: false)
         actionHandlersByProduct[keyWithRemoteProductID] = handler
+
+        updateProductIDOfImagesUploadedUsingLocalProductID(siteID: siteID,
```
❓ since there could be images pending upload at the time of this call, those images in progress might still be missing a parent ID (as mentioned in the test result in the review summary). Does it make sense to call this in `saveProductImagesWhenNoneIsPendingUploadAnymore` when there are no images pending upload?

selanthiraiyan replied:
Great point. 
I have updated the code to call `saveProductImagesWhenNoneIsPendingUploadAnymore` after saving product.

jaclync reviewed (commented):
The testing steps worked great, but I wonder if we want to support this scenario where the user saves a new product before images are uploaded:

- Launch the app
- Navigate to `Products` tab and tap on `+` button on top right.
- In product creation form enter a product title and add an image.
- ~~Wait for `image A` to finish uploading.~~
- Tap "Publish" **immediately** to save the product. 
- Stay or leave the product detail screen.
- From your computer, open your store's `Media Library` by navigating to https://yourstore.com/wp-admin/upload.php
- In the `Media Library` page,
  - Tap on the uploaded image and verify that `Uploaded to:` field is available. [Ref screenshot](https://user-images.githubusercontent.com/524475/175935720-22ba1b22-365b-49e5-95a1-59522dd9a58f.png) You can tap on the link to open the product that you created above.

I left a comment in `ProductImageUploader` about the timing of the call to update uploaded images. Would like to hear what you think!

selanthiraiyan committed:
Simplify `waitUntil` by removing `condition` parameter label.

selanthiraiyan committed:
Remove redundant `SessionManager` to simplify code.

selanthiraiyan committed:
Stop setting in `storesManager` global singleton as we can inject `storesManager` into `ProductImagesProductIDUpdater`.

selanthiraiyan committed:
Stop using json in MediaStoreTests and use `MockMediaRemote` instead.

selanthiraiyan committed:
Fix indentation in `MediaAction`

selanthiraiyan committed:
Update product images `parent_id` after uploading all pending images and saving product.

selanthiraiyan committed:
Merge branch 'feat/7021-show-product-update-failure-notice' into feat/7021-update-parent_id-of-images

selanthiraiyan committed:
ProductImagesProductIDUpdater
- Use `async throws` and `withCheckedThrowingContinuation` to replace `Result` usage.

selanthiraiyan committed:
Add more cases to `ProductImageUploaderError` to cover all paths in `saveProductImagesWhenNoneIsPendingUploadAnymore`. (Call completion handler in all paths)

selanthiraiyan committed:
Remove whitespace.

selanthiraiyan committed:
Merge branch 'trunk' into feat/7021-update-parent_id-of-images

selanthiraiyan committed:
Remove unused error cases from `ProductImageUploaderError`

selanthiraiyan committed:
Remove unused extra cases from `ProductImageUploaderError`.

selanthiraiyan committed:
Call `onProductSave` only when product is saved.

selanthiraiyan committed:
ProductImageUploaderTests
- Test that product ID of images are updated upon saving product.

selanthiraiyan committed:
Readd `test_calling_replaceLocalID_updates_excluded_product_from_status_updates` method.

selanthiraiyan commented:
Thank you for the review,  @jaclync. 🙇 

I have made the suggested changes and also [raised a different PR to cover the JCP sites](https://github.com/woocommerce/woocommerce-ios/pull/7167) to avoid bloating this PR. 

This PR is ready for another look now.

selanthiraiyan committed:
Fix grammar.

selanthiraiyan committed:
Remove whitespace.

jaclync reviewed (approved):
Thanks for the updates, LGTM! 🚀

selanthiraiyan committed:
Merge branch 'trunk' into feat/7021-update-parent_id-of-images
